src/marketplace.js

```diff
@@ -0,0 +1,321 @@
+     * 
+     * @param {string} vendorId the vendor's public key multihash
+     * @returns {Promise<ECPair>}
+     * @memberof Marketplace
+     */
+    async getVMKeyPair(vendorId) {
+        await this.start();
+        const wif = await this.db.getVMKeyPairWIF(vendorId);
+        return ECPair.fromWIF(wif);
+    }
+
+    async registerVendor(vendorPubKeyMultihash) {
+        const id = vendorPubKeyMultihash;
+        await this.start();
+        const vmKeyPair = ECPair.makeRandom();
+        const pubKeyBuffer = vmKeyPair.getPublicKeyBuffer();
+        const vmPubKeyMultihash = await this.chluIpfs.instance.vendor.storePublicKey(pubKeyBuffer);
+        this.chluIpfs.pin(vmPubKeyMultihash);
+        this.chluIpfs.pin(vendorPubKeyMultihash);
+        // TODO: request pin?
+        const keys = await this.getKeys();
+        const signature = await this.chluIpfs.instance.vendor.signMultihash(vmPubKeyMultihash, keys.keyPair);
+        const vendor = await this.db.createVendor(id, {
+            vmKeyPairWIF: vmKeyPair.toWIF(),
+            vmPubKeyMultihash,
+            mSignature: signature,
+            vSignature: null,
+            vPubKeyMultihash: id
+        });
+        const response = {
+            vPubKeyMultihash: vendor.vPubKeyMultihash,
+            vmPubKeyMultihash: vendor.vmPubKeyMultihash,
+            mSignature: vendor.mSignature 
+        };
+        return response;
+    }
+
+    /**
+     * @typedef {Object} PoPR
+     * @param {string} item_id
+     * @property {string} invoice_id
+     * @property {string} customer_id
+     * @property {integer} created_at
+     * @property {integer} expires_at
+     * @property {string} currency_symbol
+     * @property {integer} amount
+     * @property {string} marketplace_url
+     * @property {string} marketplace_vendor_url
+     * @property {string} key_location
+     * @property {string} hash calculated automatically
+     * @property {string} signature applied automatically
+     */
+
+    /**
+     * Create/Update the vendor signature for a vendor-marketplace key
+     * 
+     * @param {string} vendorId the vendor's public key multihash
+     * @returns {Promise}
+     * @throws {Error} if the signature is not valid or the vendor does not exist
+     * @memberof Marketplace
+     */
+    async updateVendorSignature(vendorPubKeyMultihash, signature) {
+        const id = vendorPubKeyMultihash;
+        await this.start();
+        const vendor = await this.db.getVendor(id);
+        if (vendor) {
+            // TODO: signature needs expiration date?
+            const PvmMultihash = vendor.vmPubKeyMultihash;
+            const valid = this.chluIpfs.instance.vendor.verifyMultihash(vendorPubKeyMultihash, PvmMultihash, signature);
+            if (valid) {
+                vendor.vSignature = signature;
+                await this.db.updateVendor(id, vendor);
+            } else {
+                throw new Error('Signature is not valid');
```

Comment:
Throw exceptions which can be understood by express app to return meaningful responses.